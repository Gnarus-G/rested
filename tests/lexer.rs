use insta::assert_debug_snapshot;

macro_rules! assert_lexes {
    ($input:literal) => {
        let lexer = rested::lexer::Lexer::new($input);

        insta::with_settings!({
             description => $input
        }, {
            assert_debug_snapshot!(lexer.into_iter().collect::<Vec<rested::lexer::Token>>());
        })
    }
}

#[test]
fn lex_put_patch_delete() {
    assert_lexes!(
        r#"
put /api {}
patch /api {}
delete /api {}
"#
    );
}

#[test]
fn lex_string_literals() {
    assert_lexes!(r#""hello""#);

    assert_lexes!(r#""hello"#);

    assert_lexes!(
        r#"
"hello
"world
"#
    );

    assert_lexes!(r#" "" "" ``"#);

    assert_lexes!(r#" { "Bearer token" } "#);

    assert_lexes!(
        r#"`
{
    stuff
}`

`
stuff"#
    );
}

#[test]
fn lex_bools() {
    assert_lexes!("true false");
}

#[test]
fn lex_numbers() {
    assert_lexes!("123124 1.0");
}

#[test]
fn lex_get_url() {
    assert_lexes!("get http://localhost");
}

#[test]
fn lex_get_url_with_header() {
    assert_lexes!("get http://localhost { header \"Authorization\" \"Bearer token\" }");
}

#[test]
fn lex_get_url_over_many_lines() {
    assert_lexes!("get\nhttp://localhost");

    assert_lexes!(
        r#"get 
    http://localhost 
{
}"#
    );
}

#[test]
fn lex_get_url_with_header_and_body() {
    assert_lexes!(
        r#"
post http://localhost { 
    header "Authorization" "Bearer token" 
    body "{neet: 1337}" 
}"#
    );
}

#[test]
fn lex_call_expression() {
    assert_lexes!(r#"env() env("stuff")"#);
}

#[test]
fn lex_template_literals() {
    assert_lexes!(r#"`stuff${"interpolated"}(things${env("dead_night")}` `dohickeys`"#);

    assert_lexes!(r#"`a${"temp"}` }}"#);

    assert_lexes!(
        r#"
let b = `asdf ${}`
"#
    );

    assert_lexes!(r#"` ${} ${} `"#);
    assert_lexes!(r#"`head ${} ${} tail`"#);

    assert_lexes!(
        r#"
let b = `asdf ${test} ${} ${"word"}`
"#
    );

    assert_lexes!(
        r#"
let b = `asdf ${}` 
let c = {}
"#
    );

    assert_lexes!(r#"`${test} ${df} `"#);
    assert_lexes!(r#"`asdf ${test} ${df} `"#);
}

#[test]
fn lex_json_object() {
    assert_lexes!(
        r#"
let obj = {
    key: "value",
    otherKey: ["val", "vlue"]
}
"#
    );
}
