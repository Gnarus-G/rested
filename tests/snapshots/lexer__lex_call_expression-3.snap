---
source: tests/lexer.rs
description: "get `http://localhost:8080/api?sort=${sort}&filter=${filter}`\nlet a = read(\"testasdf.rd\")"
expression: "lexer.into_iter().collect::<Vec<rested::lexer::Token>>()"
---
[
    Get("get") at Position { value: 0, line: 0, col: 0 },
    OpeningBackTick("`") at Position { value: 4, line: 0, col: 4 },
    StringLiteral("http://localhost:8080/api?sort=") at Position { value: 5, line: 0, col: 5 },
    DollarSignLBracket("${") at Position { value: 36, line: 0, col: 36 },
    Ident("sort") at Position { value: 38, line: 0, col: 38 },
    RBracket("}") at Position { value: 42, line: 0, col: 42 },
    StringLiteral("&filter=") at Position { value: 43, line: 0, col: 43 },
    DollarSignLBracket("${") at Position { value: 51, line: 0, col: 51 },
    Ident("filter") at Position { value: 53, line: 0, col: 53 },
    RBracket("}") at Position { value: 59, line: 0, col: 59 },
    ClosingBackTick("`") at Position { value: 60, line: 0, col: 60 },
    Let("let") at Position { value: 62, line: 1, col: 0 },
    Ident("a") at Position { value: 66, line: 1, col: 4 },
    Assign("=") at Position { value: 68, line: 1, col: 6 },
    Ident("read") at Position { value: 70, line: 1, col: 8 },
    LParen("(") at Position { value: 74, line: 1, col: 12 },
    StringLiteral("\"testasdf.rd\"") at Position { value: 75, line: 1, col: 13 },
    RParen(")") at Position { value: 88, line: 1, col: 26 },
]
